Instruction-tuning Support (#196)
Modalities/modalities: SFT_README.md

# Supervised Fine-tuning with Modalities

Currently supported are Instruction-tuning and Low-rank Adaption (LorA), as explained in more detail next.

## Instruction-tuning
* entry point to prepare data
* jinja2 templates
* The b_include_to_loss_token, e_include_to_loss_token are required to be part of each chat template for proper loss masking!
* hash to connect files
 
* truncation, padding

> * re-use last target

Reviewer: these bullet points come out of the blue

### Create Prompts from Conversations

> To prepare the instruction-tuning data we created a new entry point `apply_chat_template`, which requires a [configuration file](./config_files/data_preparation/apply_chat_template_config.yaml). Wihtin it we define:

Reviewer: ```suggestion
To prepare the instruction-tuning data we created a new entry point `apply_chat_template`, which requires a [configuration file](./config_files/data_preparation/apply_chat_template_config.yaml). Within it we define:
```

> * the path to instruction-tuning dataset as a JSONL file wereas each line contains a structured conversation as an array of dictionaries.

Reviewer: ```suggestion
* the path to instruction-tuning dataset as a JSONL file, in which each line contains a structured conversation as an array of dictionaries.
```

Reviewer: Example json line would be great

* A [jinja2](https://jinja.palletsprojects.com/en/3.1.x/) chat template which defines the rules how to glue `chat_template_data` and the data within the JSONL together to one `chat` string.

As part of the `chat_template_data`, we require the special tokens `b_include_to_loss_token` and `e_include_to_loss_token`. 
> ❗ You should choose sequences which are tokenized into a single token and will not appear in the assistant utterances of the instruction-tuning data!

They are used to mark the begin and end of the assistant turns, as we need to include only tokens between those into the loss computation during instruction-tuning with modalities.

```yaml
chat_template_data:
  ...
  special_tokens:
      b_include_to_loss_token: ^
      e_include_to_loss_token: $

> ```

Reviewer: Example line would be nice with an explanation on what parts we would consider the loss only

Run the `apply_chat_template` entry point with:
```bash
modalities data apply_chat_template --config_file_path config_files/data_preparation/apply_chat_template_config.yaml
```

This will create two files

> 1. The new JSONL file with a new attribute `chat` containing the conversations e.g. `lorem_ipsum_sft_converted.aadd295.jsonl`

Reviewer: If I understood correctly, the new JSONL file is a copy of the input JSONL file, extended with the "chat" key word. Also here an example would be nice.

2. The config used to generate the `chat` e.g. `sft_chat_template_config.aadd295.yaml`

> Both files names contain the first 7 symbols of the hash of the config file, to group files which belong together!

### Create idx and pbin files

> Before continuing with the instruction-tuning you need to index the created JSONL and convert it to a packed data file.

Reviewer: It's not really "packed" (only tokenized and padded sequences right?) and we already have the misnomer in other parts of the code. Maybe say tokenized binary file?

* We need a tokenizer to tokenize the `b_include_to_loss_token` and `e_include_to_loss_token`
* We need to not re-use the last token

For example (Copied from [config_files/training/config_lorem_ipsum_sft.yaml](config_files/training/config_lorem_ipsum_sft.yaml)):
```yaml
collate_fn:  
  component_key: collate_fn
  variant_key: mask_loss_collator_wrapper
  config:
    wrapped_collate_fn:  
      component_key: collate_fn
      variant_key: gpt_2_llm_collator
      config:
        sample_key: ${settings.referencing_keys.sample_key}
        target_key: ${settings.referencing_keys.target_key}
    target_keys_to_mask:
      - ${settings.referencing_keys.target_key}
    loss_ignore_index: -100
    mask_tokens:
      b_include_to_loss_token: ^
      e_include_to_loss_token: $
    tokenizer:
      instance_key: tokenizer
      pass_type: BY_REFERENCE
```
and
```yaml
train_dataset:
  component_key: dataset
  variant_key: packed_mem_map_dataset_continuous
  config:
    raw_data_path: ./data/lorem_ipsum_sft_converted.aadd295.pbin
    sequence_length: ${settings.training.sequence_length}
    sample_key:  ${settings.referencing_keys.sample_key}
    reuse_last_target: true
```

> # TODO

Reviewer: open todo here

    target_keys_to_mask:
      - ${settings.referencing_keys.target_key}
    loss_ignore_index: -100
    mask_tokens:
      b_include_to_loss_token: ^
      e_include_to_loss_token: $
    tokenizer:
      instance_key: tokenizer
      pass_type: BY_REFERENCE
```

with

```yaml
train_dataset:
  component_key: dataset
  variant_key: packed_mem_map_dataset_continuous
  config:
    raw_data_path: ./data/lorem_ipsum_sft_40e0699/lorem_ipsum_sft_converted.40e0699.pbin
    sequence_length: ${settings.training.sequence_length}
    sample_key:  ${settings.referencing_keys.sample_key}
    reuse_last_target: true
```

and with

```yaml
tokenizer:
  component_key: tokenizer
  variant_key: pretrained_hf_tokenizer
  config:
    pretrained_model_name_or_path: data/tokenizer/hf_gpt2
    padding: false
    truncation: false
    special_tokens:
      additional_special_tokens: 
        - "^"
        - "$"
```

Finally, run the instruction-tuning with the `run` entry point:
```bash
torch.distributed.run --nnodes 1 --nproc_per_node 2 --rdzv-endpoint=0.0.0.0:29555 src/modalities/__main__.py run --config_file_path config_files/training/config_lorem_ipsum_sft.yaml
```

> > Note, that it is advised to add a special token (which is already known as non-special token to the tokenizers' voabulary) to indicate the end of an assistant turn within the `b_include_to_loss_token` and `e_include_to_loss_token` in your chat template. Change your chat template accordingly and make sure to inlcude this token as special token in the tokenizer configuration for the pbin file creation step and model training!

Reviewer: ```suggestion
> Note, that it is advised to add a special token (which is already known as non-special token to the tokenizer's voabulary) to indicate the end of an assistant turn within the `b_include_to_loss_token` and `e_include_to_loss_token` in your chat template. Change your chat template accordingly and make sure to inlcude this token as special token in the tokenizer configuration for the pbin file creation step and model training!
```

#### A Note on Tokanization in Huggingface

> The special tokens are added to a [Trie](https://en.wikipedia.org/wiki/Trie), so that longer special tokens are split first and then shorter special tokens.

Reviewer: > and then shorter special tokens.

This does not really make sense to me, also from a linguistic point of view. Maybe rewrite it?

Example from the huggingface documentation:

```python
>>> trie = Trie()
>>> trie.split("[CLS] This is a extra_id_100")
["[CLS] This is a extra_id_100"]

>>> trie.add("[CLS]")
>>> trie.add("extra_id_1")
>>> trie.add("extra_id_100")
>>> trie.split("[CLS] This is a extra_id_100")
["[CLS]", " This is a ", "extra_id_100"]
```

When we add a special token, which exists within the tokenizer voabulary already, HF only marks it as special token (adds it to the trie).

> This means, if the sequence we add as special token already exists in the vocab, there is no need to resize the embedding matrix!

Reviewer: so basically we reflect the special token by at least on of the original tokens in the vocabulary? 

For example, assuming "spec_tok_1" is supposed to be a special token, it gets tokenized to e.g.,  `["spec", "_", "tok", "_", "1"]` ? Since it has been added to the Trie, we therefore still know that it is special token but in training the special token would still be represented by a 5 token sequence and therefore the embedding matrix can remain unchanged? 

What happens if the special token cannot be represented?

Author: Not quite. 
Given `spec_tok_1` as a special token, which is part of the trie and given the text to tokenize being `Hi there spec_tok_1!`:
First, the split method of the trie is used to extract all special tokens. In our case, the split would look like this: `["Hi there", "spec_tok_1", "!"]`, then all strings that are not part of the trie, i.e., the normal tokens, are tokenized leading to e.g. `["Hi", "there", "spec_tok_1", "!"]`
